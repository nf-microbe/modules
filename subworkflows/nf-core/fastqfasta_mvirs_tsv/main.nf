// Import modules
include { MVIRS_INDEX   } from '../../../modules/nf-core/mvirs/index'
include { MVIRS_OPRS    } from '../../../modules/nf-core/mvirs/oprs'
include { MVIRS_PARSER  } from '../../../modules/nf-core/mvirs/parser'

// Import subworkflows
include { rmEmptyFastAs; rmEmptyFastQs  } from '../utils_nfmicrobe_functions'

workflow FASTQFASTA_MVIRS_TSV {

    take:
    fastq_gz            // channel: [ [ meta.id, meat.group, meta.single_end ], [ path(fastq_1), path(fastq_2) ] ]
    fasta_gz            // channel: [ [ meta.id ], path(fasta)]
    integrase_hmm       // path: "assets/hmms/integrases/integrases.hmm"

    main:

    ch_versions = Channel.empty()

    //
    // MODULE: Index reference genome/assembly
    //
    MVIRS_INDEX(
        fasta_gz
    )
    ch_versions = ch_versions.mix(MVIRS_INDEX.out.versions)

    // join fastQ and Fasta
    ch_mvirs_oprs_input = fastq_gz.map { meta, fastq -> [ [ id:meta.group ], meta, fastq ] }
        .combine(fasta_gz, by:0)
        .combine(MVIRS_INDEX.out.index, by:0)
        .multiMap { meta_group, meta_fastq, fastq, fasta, index ->
            fastq:  [ meta_fastq, fastq ]
            fasta:  [ meta_fastq, fasta ]
            index:  [ meta_fastq, index ]
        }

    //
    // MODULE: Align reads to reference genome/assembly
    //
    MVIRS_OPRS(
        ch_mvirs_oprs_input.fastq,
        ch_mvirs_oprs_input.fasta,
        ch_mvirs_oprs_input.index
    )
    ch_versions = ch_versions.mix(MVIRS_OPRS.out.versions)

    // remove channels with empty proviruses
    ch_mvirs_provirus_fasta_gz  = rmEmptyFastAs(MVIRS_OPRS.out.fasta, false)

    // join fasta and mvirs prophages
    ch_mvirs_drs_input = ch_mvirs_provirus_fasta_gz.map { meta, fastq -> [ [ id:meta.group ], meta, fastq ] }
        .combine(fasta_gz, by:0)
        .multiMap { meta_group, meta, provirus, contigs ->
            contigs:  [ meta, contigs ]
            provirus: [ meta, provirus ]
        }

    //
    // MODULE: Detect direct repeats in proviruses
    //
    MVIRS_PARSER(
        ch_mvirs_drs_input.contigs,
        ch_mvirs_drs_input.provirus,
        integrase_hmm
    )
    ch_versions = ch_versions.mix(MVIRS_PARSER.out.versions)

    emit:
    mvirs_prophages = MVIRS_OPRS.out.prophage       // channel: [ [ meta ], prophage.tsv ]
    mvirs_fasta     = MVIRS_OPRS.out.fasta          // channel: [ [ meta ], provirus.fasta.gz ]
    mvirs_clipped   = MVIRS_OPRS.out.clipped        // channel: [ [ meta ], clipped.tsv ]
    mvirs_oprs      = MVIRS_OPRS.out.oprs           // channel: [ [ meta ], oprs.tsv ]
    mvirs_summary   = MVIRS_PARSER.out.summary      // channel: [ [ meta ], summary.tsv ]
    mvirs_integrases= MVIRS_PARSER.out.integrases   // channel: [ [ meta ], integrases.tsv ]
    versions        = ch_versions                   // channel: [ versions.yml ]
}
