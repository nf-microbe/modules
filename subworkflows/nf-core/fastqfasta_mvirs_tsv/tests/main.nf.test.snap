{
    "ERR4552622 - fastq.gz - fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.summary.tsv:md5,25cc97209f438b854fbd3093175012ee"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.summary.tsv:md5,25cc97209f438b854fbd3093175012ee"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.integrases.tsv:md5,94dc3d9060d89c6cb895dd9d581f73d4"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.integrases.tsv:md5,94dc3d9060d89c6cb895dd9d581f73d4"
                    ]
                ],
                "6": [
                    "versions.yml:md5,1efbf072e194483bdbd2dbdc3ea3ad0d",
                    "versions.yml:md5,1efbf072e194483bdbd2dbdc3ea3ad0d",
                    "versions.yml:md5,2bbd7571a3157e4cef36f8e1a5b13d81",
                    "versions.yml:md5,2bbd7571a3157e4cef36f8e1a5b13d81",
                    "versions.yml:md5,676f25ae084db4cbe3d8ae446ec39372"
                ],
                "mvirs_clipped": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ]
                ],
                "mvirs_fasta": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ]
                ],
                "mvirs_integrases": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.integrases.tsv:md5,94dc3d9060d89c6cb895dd9d581f73d4"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.integrases.tsv:md5,94dc3d9060d89c6cb895dd9d581f73d4"
                    ]
                ],
                "mvirs_oprs": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ]
                ],
                "mvirs_prophages": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ]
                ],
                "mvirs_summary": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.mvirs.summary.tsv:md5,25cc97209f438b854fbd3093175012ee"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.mvirs.summary.tsv:md5,25cc97209f438b854fbd3093175012ee"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1efbf072e194483bdbd2dbdc3ea3ad0d",
                    "versions.yml:md5,1efbf072e194483bdbd2dbdc3ea3ad0d",
                    "versions.yml:md5,2bbd7571a3157e4cef36f8e1a5b13d81",
                    "versions.yml:md5,2bbd7571a3157e4cef36f8e1a5b13d81",
                    "versions.yml:md5,676f25ae084db4cbe3d8ae446ec39372"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T11:15:07.362586371"
    }
}